File: Tutorials/ObjC/Example15/Example15/BeaconsViewController.m
//
//  BeaconsViewController.m
//  Example15
//
//  Created by Penthera on 07/22/20.
//  Copyright © 2020 penthera. All rights reserved.
//

#import "BeaconsViewController.h"

@interface BeaconsViewController ()

@property (nonatomic, strong) NSMutableArray* beaconSections;
@property (nonatomic, strong) NSMutableDictionary* beaconRowsBySection;

@end

@implementation BeaconsViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    self.navigationItem.title = @"Beacons";
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    
    if (self.beaconSections) {
        return [self.beaconSections count];
    }
    
    return 0;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    
    if (self.beaconRowsBySection)
    {
        NSArray* beaconRows = [self.beaconRowsBySection objectForKey:@(section)];
        
        if (beaconRows)
        {
            return [beaconRows count];
        }
    }
    
    return 0;
}

- (nullable NSString*)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
     
    if (section < [self.beaconSections count])
    {
        return [self.beaconSections objectAtIndex:section];
    }
    
    return nil;
}

-(CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section
{
    UILabel* headerLabel = [[UILabel alloc] init];

    headerLabel.font = [UIFont boldSystemFontOfSize:10.0];
    headerLabel.text = [self tableView:tableView titleForHeaderInSection:section];
    headerLabel.numberOfLines = 0;
    headerLabel.lineBreakMode = NSLineBreakByWordWrapping;
    headerLabel.backgroundColor = [UIColor groupTableViewBackgroundColor];
    headerLabel.textAlignment = NSTextAlignmentCenter;
    
    [headerLabel sizeToFit];

    return headerLabel.frame.size.height;
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section
{
    UILabel* headerLabel = [[UILabel alloc] init];
    
    headerLabel.font = [UIFont boldSystemFontOfSize:10.0];
    headerLabel.text = [self tableView:tableView titleForHeaderInSection:section];
    headerLabel.numberOfLines = 0;
    headerLabel.lineBreakMode = NSLineBreakByWordWrapping;
    headerLabel.backgroundColor = [UIColor groupTableViewBackgroundColor];
    headerLabel.textAlignment = NSTextAlignmentCenter;

    [headerLabel sizeToFit];

    return headerLabel;
}

#pragma mark - Table view data source

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString* cellIdentifier = @"BeaconCell";
    
    UITableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    
    if( cell == nil )
    {
        cell = [[UITableViewCell alloc]initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:cellIdentifier];
    }
    
    [cell.textLabel setFont:[UIFont fontWithName:@"Menlo" size:8.0]];
    [cell.detailTextLabel setFont:[UIFont fontWithName:@"Menlo" size:8.0]];
    cell.detailTextLabel.numberOfLines = 0;
    
    cell.textLabel.text = @"";
    
    NSArray* beaconRows = [self.beaconRowsBySection objectForKey:@(indexPath.section)];
          
    if (beaconRows)
    {
        if (indexPath.row < [beaconRows count])
        {
            NSString* beaconInfo = [beaconRows objectAtIndex:indexPath.row];
           
            if (beaconInfo)
            {
                cell.detailTextLabel.text = beaconInfo;
            }
        }
    }
    
    return cell;
}

#pragma mark - Set up beacon data to display

-(void)setupBeacons:(NSArray*)beaconsReported
{
    self.beaconSections = [NSMutableArray new];
    self.beaconRowsBySection = [NSMutableDictionary new];
        
    unsigned int currentSection = 0;
    
    NSNumber* lastAdInstance = @(0);
    NSString* lastAdUUID = @"";
    
    NSMutableArray* currentBeacons = nil;
   
    for (NSUInteger i = 0; i < [beaconsReported count]; i++)
    {
        NSDictionary* beaconInfo = [beaconsReported objectAtIndex:i];
        
        if (beaconInfo)
        {
            // Sections will demarcated by ad and instance
            
            NSNumber* currentAdInstance = [beaconInfo objectForKey:@"ads_instance"];
            NSString* currentAdUUID = [beaconInfo objectForKey:@"ads_UUID"];
            
            if (currentAdUUID && currentAdInstance)
            {
                if (![currentAdInstance isEqualToNumber:lastAdInstance] || ![currentAdUUID isEqualToString:lastAdUUID])
                {
                    // New section
                    
                    if (currentBeacons)
                    {
                        [self.beaconRowsBySection setObject:currentBeacons forKey:@(currentSection++)];
                    }
                    
                    // Section info
                                   
                    NSString* adName = [beaconInfo objectForKey:@"ads_name"];
                    NSNumber* adBreak = [beaconInfo objectForKey:@"ads_break"];
                    NSString* adDurationSeconds = [beaconInfo objectForKey:@"ads_duration_seconds"];
                    
                    NSString* headerText = [NSString stringWithFormat:@"%@ - %@ seconds - Break %@ - Instance %@", adName, adDurationSeconds, adBreak, currentAdInstance];

                    [self.beaconSections addObject:headerText];
                                        
                    currentBeacons = [NSMutableArray new];
                }
                
                // Row info
                
                NSString* adEventName = [beaconInfo objectForKey:@"ads_event_name"];
                NSString* beaconData = [beaconInfo objectForKey:@"ads_beacon_data"];
                NSNumber* httpResponseCode = [beaconInfo objectForKey:@"httpResponseCode"];
                NSNumber* adTimeOffset = [beaconInfo objectForKey:@"ads_time_offset"];
                NSString* url = [beaconInfo objectForKey:@"url"];
                
                NSMutableString* detailText = [NSMutableString new];
                
                [detailText appendFormat:@"      Event: %@\n", adEventName];
                [detailText appendFormat:@"     Beacon: %@\n", beaconData];
                [detailText appendFormat:@"Time Offset: %@\n", adTimeOffset];
                [detailText appendFormat:@"   Response: %@\n", httpResponseCode];
                [detailText appendFormat:@"        URL: %@", url];
                                
                [currentBeacons addObject:detailText];
                
                lastAdInstance = currentAdInstance;
                lastAdUUID = currentAdUUID;
            }
        }
    }
    
    if (currentBeacons)
    {
        [self.beaconRowsBySection setObject:currentBeacons forKey:@(currentSection)];
    }
        
    return;
}

@end

File: Tutorials/Swift/Example15/Example15/BeaconsViewController.swift
//
//  BeaconsViewController.swift
//  Example15
//
//  Created by dev on 7/24/20.
//  Copyright © 2020 Penthera. All rights reserved.
//

import UIKit

class BeaconsViewController: UITableViewController {

    var beaconSections: [String] = []
    var beaconRowsBySection: [UInt:[String]] = [:]
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.navigationItem.title = "Beacons"
    }

    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        return self.beaconSections.count
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        guard let beaconRows = self.beaconRowsBySection[UInt(section)] else {
            return 0
        }
        
        return beaconRows.count;
    }

    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        if (section < self.beaconSections.count)
        {
            return self.beaconSections[section]
        }
        
        return nil;
    }
    
    override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        let headerLabel = UILabel()

        headerLabel.font = UIFont.boldSystemFont(ofSize:10.0)
        headerLabel.text = self.tableView(tableView, titleForHeaderInSection: section)
        headerLabel.numberOfLines = 0
        headerLabel.lineBreakMode = .byWordWrapping
        headerLabel.backgroundColor = .groupTableViewBackground
        headerLabel.textAlignment = .center
          
        headerLabel.sizeToFit()

        return headerLabel.frame.size.height
    }
    
    override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let headerLabel = UILabel()

        headerLabel.font = UIFont.boldSystemFont(ofSize:10.0)
        headerLabel.text = self.tableView(tableView, titleForHeaderInSection: section)
        headerLabel.numberOfLines = 0
        headerLabel.lineBreakMode = .byWordWrapping
        headerLabel.backgroundColor = .groupTableViewBackground
        headerLabel.textAlignment = .center
          
        headerLabel.sizeToFit()

        return headerLabel
    }

    // MARK: - Table view data source
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: "BeaconCell") else {
            return UITableViewCell(style: .subtitle, reuseIdentifier: "BeaconCell")
        }
        
        cell.textLabel?.font = UIFont(name: "Menlo", size: 8.0)
        cell.detailTextLabel?.font = UIFont(name: "Menlo", size: 8.0)
        cell.detailTextLabel?.numberOfLines = 0
        
        cell.textLabel?.text = ""
        
        guard let beaconRows = self.beaconRowsBySection[UInt(indexPath.section)] else {
            return cell;
        }
        
        if indexPath.row < beaconRows.count
        {
            cell.detailTextLabel?.text = beaconRows[indexPath.row]
        }
        
        return cell
    }
    
     func setupBeacons(beaconsReported:[[String:Any]])
    {
        var currentSection:UInt = 0
        
        var lastAdInstance:UInt = 0
        var lastAdUUID:String = ""
        
        var currentBeacons:[String]? = nil
        
        for beaconInfo in beaconsReported
        {
            guard let currentAdInstance = beaconInfo["ads_instance"] as? UInt,
                  let currentAdUUID = beaconInfo["ads_UUID"] as? String else {
                continue;
            }
                      
            if currentAdInstance != lastAdInstance || currentAdUUID != lastAdUUID
            {
                // New section
                
                if currentBeacons != nil
                {
                    self.beaconRowsBySection[currentSection] = currentBeacons
                    
                    currentSection += 1
                }
                
                let adName = beaconInfo["ads_name"] as? String
                let adBreak = beaconInfo["ads_break"] as? UInt
                let adDurationSeconds = beaconInfo["ads_duration_seconds"] as? Double
                
                let headerText:String = "\(adName!) - \(adDurationSeconds!) seconds - Break \(adBreak!) - Instance \(currentAdInstance)"
                
                self.beaconSections.append(headerText)
                                
                currentBeacons = []
            }
            
            // Row info
            
            var detailText:String = ""
          
            let adEventName = beaconInfo["ads_event_name"] as? String
            let beaconData = beaconInfo["ads_beacon_data"] as? String
            let httpResponseCode = beaconInfo["httpResponseCode"] as? Int
            let adTimeOffset = beaconInfo["ads_time_offset"] as? Double
            let url = beaconInfo ["url"] as? String
            
            detailText += "      Event: \(adEventName!)\n"
            detailText += "     Beacon: \(beaconData!)\n"
            detailText += "Time Offset: \(adTimeOffset!)\n"
            detailText += "   Response: \(httpResponseCode!)\n"
            detailText += "        URL: \(url!)"
            
            currentBeacons?.append(detailText)
            
            lastAdInstance = currentAdInstance;
            lastAdUUID = currentAdUUID;
        }
        
        if currentBeacons != nil
        {
            self.beaconRowsBySection[currentSection] = currentBeacons
        }
        
        return
    }
}
